networkit/cpp/io/NetworkitBinaryWriter.cpp

```diff
@@ -0,0 +1,201 @@
+	Aux::enforceOpened(outfile);
+
+	bool directed = G.isDirected();
+	uint64_t format = 0;
+	unsigned weight = 0;
+	Header header = {};
+
+	auto calculateBaseOffset = [&] () {
+		return sizeof(Header);
+	};
+
+	auto calculateAdjOffset = [&] () {
+		size_t off = nodes * sizeof(uint64_t); //prefSum
+		off += chunks * sizeof(uint64_t); //first vertex
+		return sizeof(Header) + off;
+	};
+
+	auto calculateWeightsOffset = [&] () {
+		size_t off = G.numberOfEdges() * sizeof(uint8_t); //adj arrays
+		off += chunks * sizeof(uint64_t);		  // adj offsets
+		return sizeof(Header) + off + calculateAdjOffset();
+	};
+
+	auto setFeatures = [&] () {
+		header.features |= (G.isDirected() & DIR_MASK);
+		header.features |= ((weight << WGHT_SHIFT) & WGHT_MASK);
+	};
+
+	auto setHeader = [&] () {
+		strncpy(header.magic,"nkbg000",8);
+		header.checksum = 0;
+		setFeatures();
+		header.nodes = nodes;
+		header.chunks = chunks;
+		header.offsetBaseData = calculateBaseOffset();
+		header.offsetAdjLists = calculateAdjOffset();
+		header.offsetWeights = calculateWeightsOffset();
+	};
+
+	auto writeHeader = [&] () {
+
+		outfile.write(header.magic, 8);
+		outfile.write(reinterpret_cast<char*>(&header.checksum), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.features), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.nodes), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.chunks), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetBaseData), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetAdjLists), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetWeights), sizeof(uint64_t));
+	};
+
+	count prefSum = 0;
+	nodes = G.numberOfNodes();
+	if (nodes < chunks) {
+		chunks = nodes;
+		INFO("reducing chunks to ", chunks, " chunks");
+	}
+	nodesPerChunk = splitNodes(nodes, chunks);
+
+	// compute prefix sum
+	count sum = 0;
+	size_t prev = 0;
+	std::vector<count> nodePrefSum;
+	G.forNodes([&](node u) {
+		size_t adj_size = 0;
+		sum += G.degree(u);
+		nodePrefSum.push_back(sum);
+	});
+
+	// compute encoded size of arrays and store in vector
+	uint64_t prev_size = 0; 
+	uint64_t lastNode = 0;
+	std::vector<size_t> encArrSize;	//  pref sum of size encoded adj arrays
+	for(uint64_t c = 0; c < chunks; c++) {
+		uint64_t adj_size = 0;
+		for(uint64_t n = lastNode; n < (lastNode + nodesPerChunk); n++) {
+			G.forNeighborsOf(n,[&](node v) {
+				uint8_t tmp [10];
+				adj_size += encode(v, tmp);
+			});
+		}
+		adj_size += prev_size;
+		encArrSize.push_back(adj_size);
+		prev_size = adj_size;
+		lastNode = (c+1) * nodesPerChunk;
+	}
+
+	// compute chunks and their offsets
+	std::vector<std::pair<uint64_t, uint64_t>> split; // stores chunk id + first node
+	int ctr = 1;
+	int k = 0;
+	G.forNodes([&](node u) {
+		if (ctr >= nodesPerChunk) {
+			split.push_back(std::make_pair(k,(u+1)));
+			ctr = 1;
+			k++;
+		} else {
+			ctr++;
+		}
+	});
```

Comment:
AFAICS this logic does not work if the last chunk is bigger than `nodesPerChunk`.